### DISCUSSION TITLE
Would it be possible to install this as a library crate

### DISCUSSION THREAD

**Discussion by ghost** [filips123/PWAsForFirefox#352] [Q&A] [answered]
Hi :D

I've recently started my Rust journey so I don't have much experience with it, however I was wondering if it's possible to install `PWAsForFirefox` as a library through `cargo` and use the same functionality as seen in [the Native section](https://github.com/filips123/PWAsForFirefox/blob/main/native/README.md) inside a Rust project?

**Comment by filips123** [comment, marked as answer]
Do you mean just installing PWAsForFirefox with `cargo install` and using it normally (though the `firefoxpwa` program and extension), or installing it as a proper library and using it programmatically from other code? Also, if the latter, do you have any particular things that you would like to do in mind?

For the former, although it's probably possible, it's not worth it, because it would complicate installation, setup and other things, and installation through existing methods is easier.

For the latter, it's also probably possible. You can probably already add `firefoxpwa` as a Cargo Git dependency and use its public/exposed functions. However, I'm not sure if I would recommend doing this. Although the API generally won't be changing much (probably, at least for main functions), I'm not considering it as a public API, so major/breaking changes can be possible even in minor releases. Additionally, setup might be harder.

Instead, if you want to programmatically do things like installing PWAs, editing them, etc, I would recommend another approach. You can communicate with (already installed) `firefoxpwa-connector` in the same way as the browser does when using the extension, using [native messaging](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging). I don't know if there is any Rust client for native messaging yet, but the protocol is quite simple so it shouldn't be hard to implement it. Then, you could call any of [available methods](https://github.com/filips123/PWAsForFirefox/blob/main/native/src/connector/request.rs) to do various actions and get [responses](https://github.com/filips123/PWAsForFirefox/blob/main/native/src/connector/response.rs). This API can also be considered more stable, as I won't introduce breaking changes in minor releases.
